lib/spack/spack/util/compression.py

```diff
@@ -233,22 +242,23 @@ def unarchive(archive_file):
         # record name of new archive so we can extract
         # and later clean up
         decomped_tarball = decompressor(archive_file)
-        if check_extension(decomped_tarball, "tar"):
-            # run tar on newly decomped archive
-            outfile = _system_untar(decomped_tarball)
-            # clean intermediate archive to mimic end result
-            # produced by one shot decomp/extraction
-            os.remove(decomped_tarball)
-            return outfile
-        return decomped_tarball
+        # run tar on newly decomped archive
+        outfile = _system_untar(decomped_tarball)
+        # clean intermediate archive to mimic end result
+        # produced by one shot decomp/extraction
+        os.remove(decomped_tarball)
+        return outfile
 
     return unarchive
 
 
 def _py_lzma(archive_file):
     """Returns path to decompressed .xz files
     Decompress lzma compressed .xz files via python lzma module"""
-    decompressed_file = os.path.basename(strip_extension(archive_file, "xz"))
+    if check_extension(archive_file, "txz"):
+        decompressed_file = os.path.basename(strip_extension(archive_file, "txz"))
+    else:
+        decompressed_file = os.path.basename(strip_extension(archive_file, "xz"))
```

Comment:
I think this PR should apply the same level of care to `.bz2`/`.tbz` etc. I recommend making a function to streamline that:

```
def prep_archive(archive_file, extensions):
  for ext in extensions:
      outfile = strip_extension(..., ext)
      if outfile != archive_file:
        break
  else:
    outfile = archive_file
    inferred_ext = next(iter(set.union(extensions, fs.EXTS)))
    renamed_archive_file += inferred_ext
    shutil.move(archive_file, renamed_archive_file)

...

prep_archive(archive_file, ["txz", "xz"])
```

Note this also incorporates the other generalization I requested in `_system_gunzip`

Reply:
Please see my previous comment. Let me know if you prefer to do these renames consistently. We can certainly do that.

Reply:
I agree with your comment on https://github.com/spack/spack/pull/37749/files#r1258633952, but I think this particular thread raises a concern which can be considered severable, and IMO that concern persists: for example I think `_py_gunzip` should not decompress `foo.tgz` to a file that is also named `foo.tgz` (regardless of whether it can in fact do that).

Reply:
I think that's a great observation that the affected functions should have reasonable outputs even if their input has no extension. That includes tgz should not decompress to tgz. I'll look at the input and output for the affected functions on Windows and Unix.

Reply:
your concern about gunzip applied to tgz is already handled by using strip_compression_archive which expands tgz to tar.gz and then gz is stripped, so the output file is correctly infered as tar.

Reply:
I changed txz to be handled the same way.